Mathlib/Algebra/Group/Units.lean

```diff
@@ -31,29 +35,715 @@ universe u
+
+-- Porting note: unclear whether this should be a `CoeHead` or `CoeTail`
+/-- A unit can be interpreted as a term in the base `Monoid`. -/
+@[to_additive "An additive unit can be interpreted as a term in the base `AddMonoid`."]
+instance : CoeHead αˣ α :=
+  ⟨val⟩
+
+/-- The inverse of a unit in a `Monoid`. -/
+@[to_additive "The additive inverse of an additive unit in an `AddMonoid`."]
+instance : Inv αˣ :=
+  ⟨fun u => ⟨u.2, u.1, u.4, u.3⟩⟩
+
+/- porting note: the result of these definitions is syntactically equal to `Units.val` and
+`Units.inv` because of the way coercions work in Lean 4, so there is no need for these custom
+`simp` projections. -/
+#noalign units.simps.coe
+#noalign add_units.simps.coe
+#noalign units.simps.coe_inv
+#noalign add_units.simps.coe_neg
+
+-- Porting note: removed `simp` tag because of the tautology
+@[to_additive]
+theorem val_mk (a : α) (b h₁ h₂) : ↑(Units.mk a b h₁ h₂) = a :=
+  rfl
+#align units.coe_mk Units.val_mk
+#align add_units.coe_mk AddUnits.val_mk
+
+@[ext]
+theorem ext : Function.Injective (fun (u : αˣ) => (u : α))
+  | ⟨v, i₁, vi₁, iv₁⟩, ⟨v', i₂, vi₂, iv₂⟩, e => by
+    simp only at e; subst v'; congr;
+    simpa only [iv₂, vi₁, one_mul, mul_one] using mul_assoc i₂ v i₁
+#align units.ext Units.ext
+
+-- Porting note: `@[ext, to_additive]` didn't work on it's own
+attribute [to_additive] Units.ext.match_1
+attribute [to_additive] Units.ext
+#align add_units.ext AddUnits.ext
+
+@[to_additive, norm_cast]
+theorem eq_iff {a b : αˣ} : (a : α) = b ↔ a = b :=
+  ext.eq_iff
+#align units.eq_iff Units.eq_iff
+#align add_units.eq_iff AddUnits.eq_iff
+
+@[to_additive]
+theorem ext_iff {a b : αˣ} : a = b ↔ (a : α) = b :=
+  eq_iff.symm
+#align units.ext_iff Units.ext_iff
+#align add_units.ext_iff AddUnits.ext_iff
+
+/-- Units have decidable equality if the base `Monoid` has deciable equality. -/
+@[to_additive "Additive units have decidable equality
+if the base `AddMonoid` has deciable equality."]
+instance [DecidableEq α] : DecidableEq αˣ := fun _ _ => decidable_of_iff' _ ext_iff
+
+@[simp, to_additive]
+theorem mk_val (u : αˣ) (y h₁ h₂) : mk (u : α) y h₁ h₂ = u :=
+  ext rfl
+#align units.mk_coe Units.mk_val
+#align add_units.mk_coe Units.mk_val
+
+/-- Copy a unit, adjusting definition equalities. -/
+@[to_additive "Copy an `AddUnit`, adjusting definitional equalities.", simps]
+def copy (u : αˣ) (val : α) (hv : val = u) (inv : α) (hi : inv = ↑u⁻¹) : αˣ :=
+  { val, inv, inv_val := hv.symm ▸ hi.symm ▸ u.inv_val, val_inv := hv.symm ▸ hi.symm ▸ u.val_inv }
+#align units.copy Units.copy
+#align add_units.copy AddUnits.copy
+
+@[to_additive]
+theorem copy_eq (u : αˣ) (val hv inv hi) : u.copy val hv inv hi = u :=
+  ext hv
+#align units.copy_eq Units.copy_eq
+#align add_units.copy_eq AddUnits.copy_eq
+
+/-- Units of a monoid form have a multiplication and multiplicative identity. -/
+@[to_additive "Additive units of an additive monoid have an addition and an additive identity."]
+instance : MulOneClass αˣ where
```

Comment:
I think you want `attribute [instance] ...` instead; see https://github.com/leanprover-community/mathlib4/issues/667